src/main/java/org/kohsuke/github/Requester.java

```diff
@@ -497,73 +497,81 @@ public InputStream fetchStream() throws IOException {
     }
 
     private <T> T _fetch(String tailApiUrl, URL url, SupplierThrows<T, IOException> supplier) throws IOException {
-        while (true) {// loop while API rate limit is hit
+        int responseCode = -1;
+        String responseMessage = null;
+
+        int retries = CONNECTION_ERROR_RETRIES;
+
+        do {
+            // if we fail to create a connection we do not retry and we do not wrap
+            uc = null;
             uc = setupConnection(url);
 
             try {
-                return _fetchOrRetry(supplier, CONNECTION_ERROR_RETRIES);
-            } catch (IOException e) {
-                handleApiError(e);
-            } finally {
+                // This is where the request is sent and response is processing starts
+                responseCode = uc.getResponseCode();
+                responseMessage = uc.getResponseMessage();
                 noteRateLimit(tailApiUrl);
```

Comment:
We note the rate limit change as soon as we have confidence that we have enough information to do so.
Better for race conditions.